src/translator/cache.h

```diff
@@ -0,0 +1,196 @@
+#ifndef SRC_BERGAMOT_CACHE_H_
+#define SRC_BERGAMOT_CACHE_H_
+
+#include <list>
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+
+#ifndef WASM_COMPATIBLE_SOURCE
+#include "3rd_party/L4/inc/L4/LocalMemory/HashTableService.h"
+#endif
+
+#include "common/hash.h"
+#include "processed_request_sentence.h"
+#include "translator/beam_search.h"
+
+namespace marian {
+namespace bergamot {
+
+namespace cache_util {
+
+template <class HashWordType>
+struct HashWords {
+  HashWordType operator()(const Words &words) const {
+    size_t seed = 42;
+    for (auto &word : words) {
+      HashWordType hashWord = static_cast<HashWordType>(word.toWordIndex());
+      util::hash_combine<HashWordType>(seed, hashWord);
+    }
+    return seed;
+  }
+};
+
+}  // namespace cache_util
+
+struct CacheStats {
+  size_t hits{0};
+  size_t misses{0};
+  size_t activeRecords{0};
+  size_t evictedRecords{0};
+  size_t totalSize{0};
+};
+
+/// ThreadSafeL4Cache is an adapter built on top of L4 specialized for the use-case of bergamot-translator. L4 is a
+/// thread-safe cache implementation written in the "lock-free" paradigm using atomic constucts. There is locking when
+/// structures are deleted, which runs in the background using an epoch based memory reclamation (EBR) scheme. Eviction
+/// follows the "Clock" algorithm, which is a practical approximation for LRU. Records live either until they're used at
+/// least once, or if there is no space left without removing one that's not been used yet, with removal triggered to
+/// make space for new insertions.
+///
+/// This implementation creates one additional thread which manages the garbage collection for the class.  While it may
+/// be possible to get this class to compile in WASM architecture, the additional thread makes using this class
+/// unsuitable for WASM's current blocking workflow.  Keys are marian::Words, values are ProcessedRequestSentences.
+///
+/// There is a serialization (converting to binary (data*, size)) of structs overhead with this class as well, which
+/// helps keep tight lid on memory usage. This should however be cheaper in comparison to recomputing through the graph.
+
+#ifndef WASM_COMPATIBLE_SOURCE
+class ThreadSafeL4Cache {
+ public:
+  // L4 has weird interfaces with Hungarian Notation (hence IWritableHashTable). All implementations take Key and Value
+  // defined in this interface. Both Key and Value are of format: (uint8* mdata, size_t size). L4 doesn't own these - we
```

Comment:
A bit confusing. Just say "L4 makes a copy of the key and the value and stores them."

Reply:
> A bit confusing. Just say "L4 makes a copy of the key and the value and stores them."

I'm confused. As an explanation for what? 

This comment is in response to your query regarding the aliasing of types, for:

```cpp
using KeyBytes = L4::<incomprehensible-thing>
using ValueBytes = L4::<incomprehensible-thing>
```